### PR TITLE
Refactor vector storage to correctly handle relevancy scores

### PR DESCRIPTION
**Pull request by raymond-yuan**
Description: This pull request aims to support generating the correct generic relevancy scores for different vector stores by refactoring the relevance score functions and their selection in the base class and subclasses of VectorStore. This is especially relevant with VectorStores that require a distance metric upon initialization. Note many of the current implenetations of `_similarity_search_with_relevance_scores` are not technically correct, as they just return       `self.similarity_search_with_score(query, k, **kwargs)`  without applying the relevant score function

Also includes changes associated with: https://github.com/hwchase17/langchain/pull/6564 and https://github.com/hwchase17/langchain/pull/6494

See more indepth discussion in thread in #6494 

Issue: 
https://github.com/hwchase17/langchain/issues/6526
https://github.com/hwchase17/langchain/issues/6481
https://github.com/hwchase17/langchain/issues/6346

Dependencies: None

The changes include:
- Properly handling score thresholding in FAISS `similarity_search_with_score_by_vector` for the corresponding distance metric. 
- Refactoring the `_similarity_search_with_relevance_scores` method in the base class and removing it from the subclasses for incorrectly implemented subclasses.
- Adding a `_select_relevance_score_fn` method in the base class and implementing it in the subclasses to select the appropriate relevance score function based on the distance strategy.
- Updating the `__init__` methods of the subclasses to set the `relevance_score_fn` attribute.
- Removing the `_default_relevance_score_fn` function from the FAISS class and using the base class's `_euclidean_relevance_score_fn` instead.
- Adding the `DistanceStrategy` enum to the `utils.py` file and updating the imports in the vector store classes.
- Updating the tests to import the `DistanceStrategy` enum from the `utils.py` file.

### CODE REVIEW

**Comment by raymond-yuan** [comment]
@dev2049

File: langchain/vectorstores/utils.py [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
 
 import numpy as np
 
 from langchain.math_utils import cosine_similarity
 
 
+class DistanceStrategy(str, Enum):
+    EUCLIDEAN_DISTANCE = "EUCLIDEAN_DISTANCE"
+    MAX_INNER_PRODUCT = "MAX_INNER_PRODUCT"
+    DOT_PRODUCT = "DOT_PRODUCT"
```
Comment (raymond-yuan) [inline langchain/vectorstores/utils.py:14]: Note we keep DOT_PRODUCT for the singlestoredb even if most implentations the inner product, and having both is redundant..

**Comment by flepied** [comment]
Any plan to rebase this patch? I have tested it and it was working fine.

**Comment by rlancemartin** [comment]
Nice work @raymond-yuan. Can you merge master to resolve conflicts and have a look at any Lint errors (I just kicked off tests again)?

**Comment by flepied** [comment]
I have merged master at https://github.com/flepied/langchain/tree/relevance_refactor and it seems to work well with my tests...

**Comment by rlancemartin** [comment]
> I have merged master at https://github.com/flepied/langchain/tree/relevance_refactor and it seems to work well with my tests...

Great. When @raymond-yuan is back online, I'll let him resolve the merge conflicts.

I had a quick look, but there were a few to work though, so I wanted to make sure they are resolved correctly.

File: langchain/vectorstores/pinecone.py [hunk 2]
```diff
@@ -160,13 +162,27 @@ def similarity_search(
         )
         return [doc for doc, _ in docs_and_scores]
 
-    def _similarity_search_with_relevance_scores(
```
Comment (rlancemartin) [inline langchain/vectorstores/pinecone.py:163]: @raymond-yuan in the cases where `_similarity_search_with_relevance_scores` is removed (due to incorrect implementation), do we instead want to re-implement to ensure backwards compatibility?

File: langchain/vectorstores/chroma.py [hunk 3]
```diff
@@ -98,6 +109,7 @@ def __init__(
             else None,
             metadata=collection_metadata,
         )
+        self.override_relevance_score_fn = relevance_score_fn
```
Comment (rlancemartin) [inline langchain/vectorstores/chroma.py:112]: @raymond-yuan worth having a look at this related PR:
https://github.com/hwchase17/langchain/pull/7084

File: langchain/vectorstores/pinecone.py (hunk 2 above)
Comment (raymond-yuan) [reply langchain/vectorstores/pinecone.py:163]: We have a default implementation of `_similarity_search_with_relevance_scores` in the base class that will ensure backwards compatibility while also actually computing the relevance score: https://github.com/hwchase17/langchain/blob/117e9efa4fac341671c31bab892111011e034437/langchain/vectorstores/base.py#L189-L213 (not just the distance). 

All classes that previously had this ` _similarity_search_with_relevance_scores` implemented in the subclass should have a default relvance score function now that will be used

**Review by rlancemartin — APPROVED** [review]
lgmt, nice add @raymond-yuan! agree, cleaner to move to base class where consistency (0-1 scoring) is enforced (whereas today it's not consistent and only some DBs, such as FAISS, implement it correctly as a score whereas [others](https://github.com/hwchase17/langchain/pull/6494) return distance). Then vectorstores can choose their own implementation of the score function.

**Comment by rlancemartin** [comment]
@raymond-yuan I see tests for chroma, faiss, pgvector, singlestore. but a larger set of DBs (pinecone, etc) are modified; tests for some like pinecone are tricky b/c they are no in memory, but do we have confirmation (e.g., from users) that this is a no-op / backwards compatible? i see the the existing logic is typically moved to `_select_relevance_score_fn`, so it **should** be, but would be nice to verify it we could (e.g., a subtle change in scoring may confuse folks).

**Comment by rlancemartin** [comment]
> relevancy score function already implemented, these should be no-ops and completely backwards compatible

Ya, agreed. Cool, thanks for running some additional tests.

Going to merge this now, esp given the demand to get it in.

We can take up follow-up work on this in newer PRs.

Good work!

**Comment by ShantanuNair** [comment]
@raymond-yuan Looking into this for the supabase retriever. Awesome work! Any reason why you left supabase out of this? Was it implemented correctly?

**Comment by raymond-yuan** [comment]
@ShantanuNair ah sorry, no I missed this because it already had a custom implementation of [`similarity_search_with_relevance_score`](https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/supabase.py#L138). Upon closer inspection, I believe this is correctly implemented for the default `match_document` being [used](https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/supabase.py#L333) which seems to be in line with the [documentation](https://supabase.com/blog/openai-embeddings-postgres-vector). This code already converts the default distance fn (cosine distance) into a similarity score `(1 - cos(x))`

File: langchain/vectorstores/utils.py (hunk 1 above)
Comment (jpzhangvincent) [reply langchain/vectorstores/utils.py:14]: what do you mean by "singlestoredb"?

**Comment by olegshirokikh** [comment]
Looks like relevant issue with Pinecone Cosine Similarity, which is converted back to distance - filtering out relevant documents when using `score_threshold` -- #8207

**Comment by jpzhangvincent** [comment]
I found the `score_threshold` also filters out the most relevant doc for the FAISS vectorstore. I think there are still bugs with this refactoring changes. `

First, the `FAISS.similarity_search_with_score` /`similarity_search_with_score_by_vector`  actually returns the score which already represents the similarity score instead of distance... (I validated with the `embed_function` with the same sentences and the dot product is 1 which matches the score from the function) but somehow the docstring says the lower score means more similarity? 

Second, the `_max_inner_product_relevance_score_fn`  should just return the score itself for FAISS (at least) if the higher score from `FAISS.similarity_search_with_score` /`similarity_search_with_score_by_vector` represents the higher similarity... let me know if I miss anything.   

I guess a fix is to implement the `_max_inner_product_relevance_score_fn` in the `FAISS` vectorstore file, like `lambda x: x`.

**Pull request merged**